mantidproject/mantid
# Add algorithm for creating POLREF transmission workspaces

### Description of work

#### Summary of work


A new algorithm that implements the pre-processing steps described on the linked issue. See the "_**Calibration runs**_" section for a description of what is required. The purpose of the algorithm is to pre-process the magnetic and non-magnetic calibration runs taken on POLREF so that they are ready to pass to the `PolarizationEfficienciesWildes` algorithm.



Fixes #35682. 




#### Further detail of work

This PR includes a few workarounds in order to prevent delays to the work. However, ideally more fundamental fixes should be found, so I will open issues for the following:

- It seems that you cannot retrieve the output workspace group from an `OutputWorkspace` property when an algorithm is run as a child on a workspace group. As explained in the comment in the code, I have implemented a workaround, but it would be good to try and resolve this at source at some point (#38473).

- The `Divide` algorithm sets the YUnit to an empty string and the distribution flag to `True` if two workspaces with the same units are divided by each other. However the Mantid  `Integration` algorithm performs the integration differently for distribution data as it assumes that it means the counts in the workspace have been divided by the bin width. The apparent discrepancy between the two definitions of "distribution" can lead to incorrect results when we create the transmission workspace after a flood correction (due to the use of the `Integration` algorithm within `CreateTransmissionWorkspaceAuto` in order to normalize by monitors). In this PR, I have manually set the flood corrected workspace back to a non-distribution as a workaround.

### To test:



This implements a workflow that the POLREF scientists have already implemented in their own code, so I have validated against their results to check that the new algorithm produces equivalent results.

The reviewer should run the new algorithm and check that the validation is robust. The following script can be used to do that:

```
out_grp = ReflectometryISISCreateTransmission(InputRuns=["POLREF32130", "POLREF32132"], ProcessingInstructions="272-290")
```

The algorithm docs should be built and checked to ensure that they are easy to understand, and there are no spelling and grammar issues.

I have run a build from branch to check that the system test passes on all operating systems, although this was a few weeks ago now so could be repeated by the reviewer if they want to.





---

### Reviewer

Please comment on the points listed below ([full description](http://developer.mantidproject.org/ReviewingAPullRequest.html)).
**Your comments will be used as part of the gatekeeper process, so please comment clearly on what you have checked during your review.** If changes are made to the PR during the review process then your final comment will be the most important for gatekeepers. In this comment you should make it clear why any earlier review is still valid, or confirm that all requested changes have been addressed.

#### Code Review

- Is the code of an acceptable quality?
- Does the code conform to the [coding standards](http://developer.mantidproject.org/Standards/)?
- Are the unit tests small and test the class in isolation?
- If there is GUI work does it follow the [GUI standards](http://developer.mantidproject.org/Standards/GUIStandards.html)?
- If there are changes in the release notes then do they describe the changes appropriately?
- Do the release notes conform to the [release notes guide](https://developer.mantidproject.org/Standards/ReleaseNotesGuide.html)?

#### Functional Tests

- Do changes function as described? Add comments below that describe the tests performed?
- Do the changes handle unexpected situations, e.g. bad input?
- Has the relevant (user and developer) documentation been added/updated?

Does everything look good? Mark the review as **Approve**. A member of `@mantidproject/gatekeepers` will take care of it.

### Gatekeeper

If you need to request changes to a PR then please add a comment and set the review status to "Request changes". This will stop the PR from showing up in the list for other gatekeepers.
